compiler/rustc_mir_transform/src/instcombine.rs

```diff
@@ -144,9 +145,51 @@ impl<'tcx> InstCombineContext<'tcx, '_> {
     }
 
     fn combine_cast(&self, _source_info: &SourceInfo, rvalue: &mut Rvalue<'tcx>) {
-        if let Rvalue::Cast(_kind, operand, ty) = rvalue {
-            if operand.ty(self.local_decls, self.tcx) == *ty {
+        if let Rvalue::Cast(kind, operand, cast_ty) = rvalue {
+            let operand_ty = operand.ty(self.local_decls, self.tcx);
+            if operand_ty == *cast_ty {
                 *rvalue = Rvalue::Use(operand.clone());
+            } else if *kind == CastKind::Transmute {
+                // Transmuting an integer to another integer is just a signedness cast
+                if let (ty::Int(int), ty::Uint(uint)) | (ty::Uint(uint), ty::Int(int)) = (operand_ty.kind(), cast_ty.kind())
+                    && int.bit_width() == uint.bit_width()
```

Comment:
The width check is also good to keep in because if/when we do allow differing sizes (without UB), this transformation is wrong on big endian targets in those cases.

Reply:
Actually, on second thought, this seems easy enough to miss that it's probably worth a comment now, to prevent a bug in the future

Reply:
Added a note about future big-endian correctness.